File: api/internal/model/analytics.go
package model

import "time"

type ClicksOverTime struct {
	Date  time.Time `json:"date"`
	Count int64     `json:"count"`
}

type DeviceBreakdown struct {
	Mobile  int64 `json:"mobile"`
	Desktop int64 `json:"desktop"`
	Tablet  int64 `json:"tablet"`
	Bot     int64 `json:"bot"`
	Unknown int64 `json:"unknown"`
}

type ReferrerBreakdown struct {
	Direct    int64 `json:"direct"`
	Instagram int64 `json:"instagram"`
	Facebook  int64 `json:"facebook"`
	Twitter   int64 `json:"twitter"`
	TikTok    int64 `json:"tiktok"`
	LinkedIn  int64 `json:"linkedin"`
	WhatsApp  int64 `json:"whatsapp"`
	YouTube   int64 `json:"youtube"`
	Google    int64 `json:"google"`
	Other     int64 `json:"other"`
}

type LinkAnalytics struct {
	LinkID       int64             `json:"link_id"`
	ShortCode    string            `json:"short_code"`
	TotalClicks  int64             `json:"total_clicks"`
	UniqueClicks int64             `json:"unique_clicks"`
	OverTime     []ClicksOverTime  `json:"over_time"`
	Devices      DeviceBreakdown   `json:"devices"`
	Referrers    ReferrerBreakdown `json:"referrers"`
}

type TopLink struct {
	ShortCode   string  `json:"short_code"`
	OriginalURL string  `json:"original_url"`
	Title       *string `json:"title"`
	TotalClicks int64   `json:"total_clicks"`
}

type ProjectAnalytics struct {
	TotalClicks  int64             `json:"total_clicks"`
	UniqueClicks int64             `json:"unique_clicks"`
	OverTime     []ClicksOverTime  `json:"over_time"`
	Devices      DeviceBreakdown   `json:"devices"`
	Referrers    ReferrerBreakdown `json:"referrers"`
	TopLinks     []TopLink         `json:"top_links"`
}

File: web/src/lib/analytics.ts
import type { DeviceBreakdown, ReferrerBreakdown, BrowserBreakdown } from './types'

export const VALID_PERIODS = ['7d', '30d', '90d'] as const
export type Period = (typeof VALID_PERIODS)[number]

export const periods: { value: Period; label: string }[] = [
	{ value: '7d', label: '7 days' },
	{ value: '30d', label: '30 days' },
	{ value: '90d', label: '90 days' },
]

export function parsePeriod(raw: string | null): Period {
	return (VALID_PERIODS as readonly string[]).includes(raw ?? '')
		? (raw as Period)
		: '30d'
}

export function deviceItems(d: DeviceBreakdown) {
	return [
		{ label: 'Mobile', value: d.mobile },
		{ label: 'Desktop', value: d.desktop },
		{ label: 'Tablet', value: d.tablet },
		{ label: 'Bot', value: d.bot },
		{ label: 'Unknown', value: d.unknown },
	]
		.filter((i) => i.value > 0)
		.sort((a, b) => b.value - a.value)
}

export function browserItems(b: BrowserBreakdown) {
	return [
		{ label: 'Chrome', value: b.chrome },
		{ label: 'Safari', value: b.safari },
		{ label: 'Firefox', value: b.firefox },
		{ label: 'Edge', value: b.edge },
		{ label: 'Opera', value: b.opera },
		{ label: 'Samsung', value: b.samsung },
		{ label: 'IE', value: b.ie },
		{ label: 'Other', value: b.other },
		{ label: 'Unknown', value: b.unknown },
	]
		.filter((i) => i.value > 0)
		.sort((a, b) => b.value - a.value)
}

export function referrerItems(r: ReferrerBreakdown) {
	return [
		{ label: 'Direct', value: r.direct },
		{ label: 'Google', value: r.google },
		{ label: 'Instagram', value: r.instagram },
		{ label: 'Facebook', value: r.facebook },
		{ label: 'Twitter/X', value: r.twitter },
		{ label: 'TikTok', value: r.tiktok },
		{ label: 'Discord', value: r.discord },
		{ label: 'LinkedIn', value: r.linkedin },
		{ label: 'WhatsApp', value: r.whatsapp },
		{ label: 'YouTube', value: r.youtube },
		{ label: 'Other', value: r.other },
	]
		.filter((i) => i.value > 0)
		.sort((a, b) => b.value - a.value)
}
